crates/jit/src/compiler.rs

```diff
@@ -51,13 +55,16 @@ pub enum CompilationStrategy {
 pub struct Compiler {
     isa: Box<dyn TargetIsa>,
 
-    code_memory: CodeMemory,
     trap_registry: TrapRegistry,
-    trampoline_park: HashMap<VMSharedSignatureIndex, *const VMFunctionBody>,
     signatures: SignatureRegistry,
     strategy: CompilationStrategy,
     cache_config: CacheConfig,
+    inner: RwLock<CompilerInner>,
```

Comment:
I personally think that this probably isn't the way we want to parallelize `Module` and `Store` and such. This means that compilation is still entirely serialized and you can't compile multiple modules in parallel, one of the main usages of enabling parallelism. I think we need to work to continue to refactor and reimplement the guts of `Compiler` to make them `Send` and `Sync` and use `&self`.

For example the `trampoline_park` field is going away in https://github.com/bytecodealliance/wasmtime/pull/957, and I don't think that the switch to `AtomicPtr<T>` here is correct in that it's not ever modified, it's just a pointer.

The main next change is to move `CodeMemory` to per-module instances, and then mutability shouldn't be needed on most of the fields here.